File: C#/ContainsNearbyDuplicate.cs
//O(N)
//O(1)
public class Solution
{
    public bool ContainsNearbyDuplicate(int[] nums, int k)
    {
        if (nums.Length <= 0 || k <= 0) return false;

        var numToIndex = new Dictionary<int, int>();

        for (int i = 0; i < nums.Length; i++)
        {
            if (numToIndex.ContainsKey(nums[i]))
            {
                int prevIndex = numToIndex[nums[i]];
                if (i - prevIndex <= k)
                {
                    return true;
                }
            }

            numToIndex[nums[i]] = i;
        }

        return false;
    }
}

File: C++/containsNearbyDuplicate.cpp
// Time: O(N)
// Space: O(N)
#include <vector>
#include <unordered_map>

class Solution
{
public:
    bool containsNearbyDuplicate(std::vector<int> &nums, int k)
    {
        if (nums.empty())
            return false;

        std::unordered_map<int, int> map;

        for (int i = 0; i < nums.size(); i++)
        {

            if (map.find(nums[i]) != map.end() && abs(i - map[nums[i]]) <= k)
            {
                return true;
            }
            map[nums[i]] = i;
        }

        return false;
    }
};